### PR TITLE
Testsuite: Add test for service pack migration

### PR DESCRIPTION
**Pull request by calancha**
Run the service pack migration for both, the minion and the
SSH managed minion at the init_clients stage.
This means we run the whole secondary features on
the migrated minions, i.e., on a newer service pack.

- testsuite/features/init_clients/minssh_sp_migration.feature
- testsuite/features/init_clients/min_sp_migration.feature:
New tests.
- testsuite/run_sets/init_clients.yml: Run the sp migration
  right after bootstrap the corresponding minion.

- testsuite/features/support/env.rb:
Add tag @service_pack_migration to enable/disable the migration.

- testsuite/features/reposync/srv_sync_products.feature
- testsuite/features/support/commonlib.rb
- testsuite/features/step_definitions/common_steps.rb:
Ensure we add both products, the one before/after the migration.
Add activation keys to support the service pack migration: one
for the normal minion and one for the SSH minion.

- testsuite/features/support/twopence_init.rb
- testsuite/features/support/constants.rb:
- testsuite/documentation/cucumber-steps.md:
Add labels 'ssh_spack_migrated_minion', 'sle_spack_migrated_minion'.

## Links
https://github.com/SUSE/spacewalk/issues/11471
This PR is a reimplementation of https://github.com/uyuni-project/uyuni/pull/3035
### Ports
- Manager-4.1: https://github.com/SUSE/spacewalk/pull/13594
- Manager-4.0: https://github.com/SUSE/spacewalk/pull/13601
 (at 4.0 only port the refactor of `compute_list_to_leave_running`)

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

### CODE REVIEW

File: testsuite/features/support/env.rb
Comment (srbarrios) [inline testsuite/features/support/env.rb:26]: In my humble opinion is less readable in that format :P

File: testsuite/features/support/env.rb
Comment (srbarrios) [inline testsuite/features/support/env.rb:327]: As far as I see the test strategy for the SP migration, we want to execute it always.
As we will deploy from the terraform file a SLES15SP1 version, and we want to test SLES15SP2 version for all our test suites features. So, I'd say is not optional to migrate to SP2.

File: testsuite/features/support/constants.rb
Comment (srbarrios) [inline testsuite/features/support/constants.rb:232]: Very nice, I love it! Thanks :)

**Review by srbarrios — APPROVED** [review]
Some thoughts about the idea of making it optional. But anyhow LGTM

File: testsuite/features/support/commonlib.rb
Comment (srbarrios) [inline testsuite/features/support/commonlib.rb:80]: Awesome

File: testsuite/features/support/constants.rb
Comment (calancha) [reply testsuite/features/support/constants.rb:232]: Haha You forgot it! It was **you** suggestion in a previous PR, so I brought here ;-)

File: testsuite/features/support/env.rb
Comment (calancha) [reply testsuite/features/support/env.rb:26]: You are absolutely right! I am going to use the previous one.

File: testsuite/features/support/env.rb
Comment (calancha) [reply testsuite/features/support/env.rb:327]: Agree. Once merged, let's keep this tag for some CI runs (passing the env variable to the pipe in order to run the SP migration) for easy comparison with/without SP migration; if we don't observe breakages, then it has sense to drop this optional tag.

**Comment by Bischoff** [comment]
For @srbarrios ' information, we discussed (Tino and I), and decided to do like this:

In `init-clients`, either run `sle_minion.feature` or `min_sp_migration.feature`, but not both. same for SSH minion. Remove the initial "unregister" too.

Goal is to accelerate `init_clients` stage.

We already do something similar for the proxy: we have 3 ways to register it, but only run one of them.

**Comment by srbarrios** [comment]
> For @srbarrios ' information, we discussed (Tino and I), and decided to do like this:
> 
> In `init-clients`, either run `sle_minion.feature` or `min_sp_migration.feature`, but not both. Same for SSH minion. Remove the initial "unregister" too.
> 
> Goal is to accelerate `init_clients` stage.
> 
> We already do something similar for the proxy: we have 3 ways to register it, but only run one of them.

@Bischoff @calancha I think better not make it optional, we should always run the sp migration feature, otherwise, we will be testing the whole test suite with sles15sp1 instead of sles15sp2.

**Comment by calancha** [comment]
> In `init-clients`, either run `sle_minion.feature` or `min_sp_migration.feature`, but not both. Same for SSH minion. Remove the initial "unregister" too.
> 
> Goal is to accelerate `init_clients` stage.

See the comment addressing this suggestion; I have tested it at head and 4.1.

File: testsuite/features/init_clients/min_sp_migration.feature
Comment (Bischoff) [inline testsuite/features/init_clients/min_sp_migration.feature:22]: When

File: testsuite/features/init_clients/min_sp_migration.feature
Comment (Bischoff) [inline testsuite/features/init_clients/min_sp_migration.feature:24]: When

File: testsuite/features/init_clients/minssh_sp_migration.feature
Comment (Bischoff) [inline testsuite/features/init_clients/minssh_sp_migration.feature:20]: When

File: testsuite/features/init_clients/minssh_sp_migration.feature
Comment (Bischoff) [inline testsuite/features/init_clients/minssh_sp_migration.feature:22]: When

File: testsuite/features/support/commonlib.rb
Comment (Bischoff) [inline testsuite/features/support/commonlib.rb:75]: No. The products for auto-installation tests should not disappear.

We need to guarantee that these products are added to the whitelist when we are testing auto-installation and that they are not when we don't.
￼
It is exactly same situation as for the service pack migration

The argument "we are going to run the service pack migration by default" is invalid, the test `  if $service_pack_migration_enabled` tells exactly the contrary.

Don't worry about putting duplicate packages in the list, the list ist reduced at the end anyway thanks to `.uniq`.

**Review by Bischoff — CHANGES_REQUESTED** [review]
A few nitpicks, but also a serious alert.

**Comment by Bischoff** [comment]
> @Bischoff @calancha I think better not make it optional, we should always run the sp migration feature, otherwise, we will be testing the whole test suite with sles15sp1 instead of sles15sp2.

@srbarrios It's possible we want to see how the testsuite reacts when using a "normal" minion  instead of a migrated minion. So there is a real use case for making it optional.

Also, I think it perfectly matches your "focus areas" strategy. What if someone just wants a quick test and is not interested into this service pack migration?

A valid roadmap would be to have these tags:
* `@scope_servicepack_migration`     (instead of current `@servicepack_migration`)
* `@scope_build_osimage` (instead of current `@long_tests`)
* `@scope_autoinstall` (instead of current `@long_tests` too)

What do you think?

File: testsuite/features/support/commonlib.rb
Comment (srbarrios) [reply testsuite/features/support/commonlib.rb:75]: I don't understand your message :)
I think nothing has been disappeared just well refactored :)

**Comment by srbarrios** [comment]
> > @Bischoff @calancha I think better not make it optional, we should always run the sp migration feature, otherwise, we will be testing the whole test suite with sles15sp1 instead of sles15sp2.
> 
> @srbarrios It's possible we want to see how the testsuite reacts when using a "normal" minion instead of a migrated minion. So there is a real use case for making it optional.
> 
> Also, I think it perfectly matches your "focus areas" strategy. What if someone just wants a quick test and is not interested into this service pack migration?
> 
> A valid roadmap would be to have these tags:
> 
> * `@scope_servicepack_migration`     (instead of current `@servicepack_migration`)
> * `@scope_build_osimage` (instead of current `@long_tests`)
> * `@scope_autoinstall` (instead of current `@long_tests` too)
> 
> What do you think?

I still think that if we don't pass the sp_migration feature, we will test the full testsuite with sles15sp1 instead of sles15sp2.
And it's hard to design a mechanism where we deploy sles15sp2 instead of sles15sp1 when a different .feature test is enabled.

Let's have an agreement, let's merge as it is, so we have the same approach in the .yaml file that we have on proxy.
BUT, let's add a warning, advising just after the sp_migration feature on the .yaml file, that if we don't run it, the rest of the test suite will run using sles15sp1.

**Comment by Bischoff** [comment]
> I still think that if we don't pass the sp_migration feature, we will test the full testsuite with sles15sp1 instead of sles15sp2.
> And it's hard to design a mechanism where we deploy sles15sp2 instead of sles15sp1 when a different .feature test is enabled.

Very true, but I think it's an acceptable problem.

> Let's have an agreement, let's merge as it is, so we have the same approach in the .yaml file that we have on proxy.
> BUT, let's add a warning, advising just after the sp_migration feature on the .yaml file, that if we don't run it, the rest of the test suite will run using sles15sp1.

Agree. Just like we do with the proxy.

For the moment, the old minion bootstrap features are commented out in the yaml file (to keep the complexity of this PR low). But on the long run we could uncomment them and proceed according to the tag to provide a real choice between "direct bootstrap" and "sp migration bootstrap".

File: testsuite/features/support/commonlib.rb
Comment (calancha) [reply testsuite/features/support/commonlib.rb:75]: OK, I see Eric's point; they are needed only at HEAD because we run such a test only at that branch.
I have just pushed a commit to reenable them into `do_not_kill`

File: testsuite/features/init_clients/min_sp_migration.feature
Comment (calancha) [reply testsuite/features/init_clients/min_sp_migration.feature]: Thank you. Done.

**Review by Bischoff — APPROVED** [review]

**Comment by Bischoff** [comment]
@calancha I approved, but if i may: the name of the array of channels to whitelist if we are testing Cedric's autoinstallation should bear a reference to this autoinstallation focus area, so we know it's in that case only.

**Comment by calancha** [comment]
> @calancha I approved, but if i may: the name of the array of channels to whitelist if we are testing Cedric's autoinstallation should bear a reference to this autoinstallation focus area, so we know it's in that case only.

Addressed with a comment at the entry of the array; I mention the purpose of such an entry.

**Pull request merged**
